src/platform_linux/laio.c

```diff
@@ -67,40 +70,140 @@ laio_cleanup(io_handle *ioh, uint64 count);
 static void
 laio_cleanup_all(io_handle *ioh);
 
+static void
+laio_register_thread(io_handle *ioh);
+
+static void
+laio_deregister_thread(io_handle *ioh);
+
 static io_async_req *
 laio_get_kth_req(laio_handle *io, uint64 k);
 
 /*
  * Define an implementation of the abstract IO Ops interface methods.
  */
 static io_ops laio_ops = {
-   .read          = laio_read,
-   .write         = laio_write,
-   .get_iovec     = laio_get_iovec,
-   .get_async_req = laio_get_async_req,
-   .get_metadata  = laio_get_metadata,
-   .read_async    = laio_read_async,
-   .write_async   = laio_write_async,
-   .cleanup       = laio_cleanup,
-   .cleanup_all   = laio_cleanup_all,
-   .get_context   = laio_get_context,
+   .read              = laio_read,
+   .write             = laio_write,
+   .get_iovec         = laio_get_iovec,
+   .get_async_req     = laio_get_async_req,
+   .get_metadata      = laio_get_metadata,
+   .read_async        = laio_read_async,
+   .write_async       = laio_write_async,
+   .cleanup           = laio_cleanup,
+   .cleanup_all       = laio_cleanup_all,
+   .register_thread   = laio_register_thread,
+   .deregister_thread = laio_deregister_thread,
+   .get_context       = laio_get_context,
+   .get_io_async_req  = laio_get_io_async_req,
 };
 
+/*
+ * Given an AIO handle, set up a thread-specific IO context opaque handle.
+ */
+static platform_status
+io_context_setup(laio_handle *io)
+{
+   int status;
+
+   const threadid tid = platform_get_tid();
+
+   // Expect that this was never setup; otherwise it's a coding error.
+   platform_assert((io->ctx[tid] == NULL),
+                   "IO-context for ThreadID=%lu is expected to be NULL"
+                   ", but it is ctx[tid]=%p\n",
+                   tid,
+                   io->ctx[tid]);
+
+   status = io_setup(io->cfg->kernel_queue_size, &io->ctx[tid]);
+   platform_assert((status == 0),
+                   "io_setup() failed for thread-ID=%lu "
+                   "with status=%d (%s)\n",
+                   tid,
+                   status,
+                   strerror(status));
+
+   return STATUS_OK;
+}
+
+/*
+ * As part of thread deregistration, we need to release the IO context
+ * that was setup for this thread. Here, we assume that required io_cleanup()
+ * has already been done to drain out pending IOs. This is quite the very
+ * last thing that happens before thread deregistration completes.
+ *
+ * This routine will also be called when Splinter is being shutdown as
+ * part of deinitializing the IO system. For that usage, allow for a case when
+ * some threads' IO-context may not have been cleaned up previously; so be
+ * a bit little lax on the assertion checks.
+ *
+ * Returns: 0, for successful destroy; 1, otherwise.
+ */
+static int
+io_context_cleanup(laio_handle *io, threadid tid, bool from_deregister)
```

Comment:
Again, needs close review.

Reply:
See comments added to the files requiring changes.

Reply:
@rtjohnso and I walked through the logic changes during our review discussion and agreed to the nature of these changes.

All comments added to clean-up code raised during this review discussion have since been addressed, with code rework.